src/mainboard/sunxi/nezha/main/src/main.rs

```diff
@@ -21,53 +21,47 @@ use spin;
 
 mod sbi_platform;
 
-const MEM: usize = 0x4000_0000;
+const RAM_BASE: usize = 0x4000_0000;
 
 // see ../fixed-dtfs.dts
-// const PAYLOAD_OFFSET: usize = 0x2_0000;
-const PAYLOAD_SIZE: usize = 0x20_0000; // 2 MB
-const PAYLOAD_ADDR: usize = MEM + 0x20_0000;
+// target location for decompressed image
+const PAYLOAD_OFFSET: usize = 0x20_0000;
+const PAYLOAD_ADDR: usize = RAM_BASE + PAYLOAD_OFFSET;
+const PAYLOAD_SIZE: usize = 0x0200_0000; // 32 MB
 
-// compressed image
-const LINUXBOOT_TMP_OFFSET: usize = 0x0400_0000;
-const LINUXBOOT_TMP_ADDR: usize = MEM + LINUXBOOT_TMP_OFFSET;
+const DTB_ADDR: usize = PAYLOAD_ADDR + PAYLOAD_SIZE;
+const DTB_SIZE: usize = 0x0001_0000;
 
-// target location for decompressed image
-const LINUXBOOT_OFFSET: usize = 0x0020_0000;
-const LINUXBOOT_ADDR: usize = MEM + LINUXBOOT_OFFSET;
-const LINUXBOOT_SIZE: usize = 0x0200_0000;
-// DTB_OFFSET should be >=LINUXBOOT_OFFSET+LINUXBOOT_SIZE and match bt0
-// TODO: Should we just copy it to a higher address before decompressing Linux?
-const DTB_OFFSET: usize = LINUXBOOT_OFFSET + LINUXBOOT_SIZE;
-const DTB_ADDR: usize = MEM + DTB_OFFSET;
+const COMPRESSED_ADDR: usize = RAM_BASE + 0x0400_0000;
+const COMPRESSED_SIZE: usize = 0x00fe_0000;
 
 fn decompress_lb() {
     // check for Device Tree header, d00dfeed
     let dtb = unsafe { read_volatile(DTB_ADDR as *const u32) };
     if dtb != 0xedfe0dd0 {
-        panic!("DTB looks wrong: {:08x}\n", dtb);
+        panic!("DTB looks wrong: {dtb:08x}");
     } else {
-        print!("DTB looks fine, yay!\n");
+        println!("DTB looks fine, yay!");
     }
 
     unsafe {
```

Comment:
So what I'd do here is something like,

        let compressed_addr = core::ptr::with_exposed_provenance::<u8>(COMPRESSED_ADDR);
        let compressed = unsafe { slice::from_raw_parts(compressed_addr, COMPRESSED_SIZE);
        let payload_addr = core::ptr::with_exposed_provenance_mut::<u8>(PAYLOAD_ADDR);
        let payload = unsafe {
            core::ptr::write_bytes(payload_addr, 0, PAYLOAD_SIZE);
            slice::from_raw_parts_mut(payload_addr, PAYLOAD_SIZE);
        };
        decompress(compressed, payload);

Reply:
Interesting, [`with_exposed_provenance()`](https://doc.rust-lang.org/stable/core/ptr/fn.with_exposed_provenance.html) is new to me - introduced in Rust 1.84.0 per the docs. I will need to read up on it in more detail. Thank you!